### PR TITLE
Launch Safari browser directly instead SafariLauncher

### PR DESCRIPTION
**Pull request by nghiadhd**
WebDriverAgent allow launching Safari browser directly
Origin appium-xcuitest-driver just launch SafariLaunch and nothing.
I fixed it by start new session with Safari browser and now it can run the test

### CODE REVIEW

File: lib/driver.js [hunk 1]
```diff
@@ -172,11 +171,11 @@ class XCUITestDriver extends BaseDriver {
 
     if (this.opts.browserName) {
       log.info('Safari test requested');
-      this.safari = true;      
+      this.safari = true;
       this.opts.app = undefined;
       this.opts.processArguments = this.opts.processArguments || {};
       if (this.isRealDevice()) {
```
Comment (imurchie) [inline lib/driver.js]: What is the need for this split, then? The only thing different, it seems, is the initial url.

File: lib/driver.js [hunk 2]
```diff
@@ -172,11 +171,11 @@ class XCUITestDriver extends BaseDriver {
...
-        this.opts.bundleId = SAFARI_LAUNCHER_BUNDLE;
+        this.opts.bundleId = SAFARI_BUNDLE_ID;
         this._currentUrl = this.opts.safariInitialUrl || `http://appium.io`;
         this.opts.processArguments.args = [this._currentUrl];
```
Comment (imurchie) [inline lib/driver.js]: If this is going straight to Safari, these process arguments need to be changed, I think.

File: lib/driver.js (hunk 1 above)
Comment (nghiadhd) [inline lib/driver.js]: It's lint issue, I just remove the white space

File: lib/driver.js (hunk 2 above)
Comment (nghiadhd) [inline lib/driver.js]: I've just followed the guides from WebDriverAgent, The bundleId is always is bundle id of Safari.
 I don't know we have plan to support for other browsers (aka chrome) or not?

```
If application is already installed on device you can start it using bundleId parameter:

curl -X POST $JSON_HEADER \
-d "{\"desiredCapabilities\":{\"bundleId\":\"com.apple.mobilesafari\"}}" \
$DEVICE_URL/session
It is also possible to request to install application before launching it by adding app parameter:

curl -X POST $JSON_HEADER \
-d "{\"desiredCapabilities\":{\"bundleId\":\"com.apple.mobilesafari\", \"app\":\"[host_path]/magicapp.app\"}}" \
$DEVICE_URL/session
```

File: lib/driver.js (hunk 1 above)
Comment (nghiadhd) [inline lib/driver.js]: OH I see, I didn't test on Simulator.
I see it should be 

```
this._currentUrl = this.opts.safariInitialUrl ? this.opts.safariInitialUrl : 'http://appium.io'
```

Comment (nghiadhd) [inline lib/driver.js]: If you still have concern about this, tomorrow I'll refactor and run long tests on my side to verify

Comment (imurchie) [inline lib/driver.js]: The current url is the only thing that is different between real device and simulator.

File: lib/driver.js (hunk 2 above)
Comment (imurchie) [inline lib/driver.js]: It's just matter of what Safari expects vs. what SafariLauncher expects (note the `-u`).

Comment (nghiadhd) [inline lib/driver.js]: In the past, I did try WebDriverAgent standalone with Simulator & Real device. I can invoke the Safari browser and test with it by this command (don't need `-u `)

```
curl -X POST $JSON_HEADER \
-d "{\"desiredCapabilities\":{\"bundleId\":\"com.apple.mobilesafari\"}}" \
$DEVICE_URL/session
```

So, tomorrow, I'll run appium with iOS 10 simulator to verify

Comment (imurchie) [inline lib/driver.js]: The `-u` allows you to specify the url to which to go.

Comment (nghiadhd) [inline lib/driver.js]: Thanks @imurchie 
I think I need to refactor a little bit for Simulator & real device

**Comment by nghiadhd** [comment]
@imurchie 

I have a question. How to trigger the CI to rerun again. I see many error issues and may it relates with configuration (ECONNREFUSE). Now I would like to run it again

**Comment by yaphatak** [comment]
@imurchie I tried this approach when ios 10 was in beta and it didnt work for me then. I will check now if this works. If it does then it would be awesome

**Comment by yaphatak** [comment]
directly launching safari works with current ios 10 version. This is really good.
@imurchie I updated my PR as well.

**Comment by yaphatak** [comment]
@imurchie You can pick this PR over mine if you want @nghiadhd : in that case please look at https://github.com/appium/appium-xcuitest-driver/pull/169 and make  necessary changes

**Comment by nghiadhd** [comment]
@imurchie 
I have just removed safari-browser as your suggestion.

File: lib/driver.js [hunk 3]
```diff
@@ -172,18 +172,12 @@ class XCUITestDriver extends BaseDriver {
...
-        this.opts.bundleId = SAFARI_LAUNCHER_BUNDLE;
-        this._currentUrl = this.opts.safariInitialUrl || `http://appium.io`;
-        this.opts.processArguments.args = [this._currentUrl];
-      } else {
-        this.opts.bundleId = SAFARI_BUNDLE_ID;
-        this._currentUrl = this.opts.safariInitialUrl || `http://${this.opts.address}:${this.opts.port}/welcome`;
-        this.opts.processArguments.args = ['-u', this._currentUrl];
-      }
+      this.opts.bundleId = SAFARI_BUNDLE_ID;
+      this._currentUrl = this.opts.safariInitialUrl || DEFAULT_SAFARI_URL;
```
Comment (imurchie) [inline lib/driver.js]: This needs to be different for simulators as real devices. It's the only part of the conditional that remains.

```js
this._currentUrl = this.opts.safariInitialUrl || 
                   this.isRealDevice() ? 'http://appium.io' : 
                   `http://${this.opts.address}:${this.opts.port}/welcome`;
```

Comment (nghiadhd) [inline lib/driver.js]: @imurchie 

I see the ops.address and opts.port are sent via cap. what happen if the end-use don't set safariInitialUrl and address/port or the url http://${this.opts.address}:${this.opts.port}/welcome doesn't exist

My suggestion we should launch just only the safari browser (or default url: http://appium.io) for both case (real device and simulator)

Comment (imurchie) [inline lib/driver.js]: We don't want to default to something served over the internet. This has caused problems in the past.

The address and port come from the server arguments when Appium is started up. They will both have values.

Comment (nghiadhd) [inline lib/driver.js]: @imurchie I got it, let's me refactor it again

**Comment by nghiadhd** [comment]
Thank @imurchie

**Pull request merged**